WalletWasabi/WabiSabi/Client/CoinJoinManager.cs

```diff
@@ -158,56 +159,51 @@ async void StartCoinJoinCommand(StartCoinJoinCommand startCommand)
 				return;
 			}
 
-			if (WalletsInSendWorkflow.ContainsKey(walletToStart.WalletName))
+			IEnumerable<SmartCoin> SanityChecksAndGetCoinCandidatesFunc()
 			{
-				ScheduleRestartAutomatically(walletToStart, trackedAutoStarts, startCommand.StopWhenAllMixed, startCommand.OverridePlebStop, stoppingToken);
+				if (WalletsInSendWorkflow.ContainsKey(walletToStart.WalletName))
+				{
+					throw new CoinJoinClientException(CoinjoinError.UserInSendWorkflow);
+				}
 
-				NotifyCoinJoinStartError(walletToStart, CoinjoinError.UserInSendWorkflow);
-				return;
-			}
+				if (walletToStart.IsUnderPlebStop && !startCommand.OverridePlebStop)
+				{
+					walletToStart.LogTrace("PlebStop preventing coinjoin.");
 
-			if (walletToStart.IsUnderPlebStop && !startCommand.OverridePlebStop)
-			{
-				walletToStart.LogTrace("PlebStop preventing coinjoin.");
-				ScheduleRestartAutomatically(walletToStart, trackedAutoStarts, startCommand.StopWhenAllMixed, startCommand.OverridePlebStop, stoppingToken);
+					throw new CoinJoinClientException(CoinjoinError.NotEnoughUnprivateBalance);
+				}
 
-				NotifyCoinJoinStartError(walletToStart, CoinjoinError.NotEnoughUnprivateBalance);
-				return;
-			}
+				if (WasabiBackendStatusProvide.LastResponse is not { } synchronizerResponse)
+				{
+					throw new CoinJoinClientException(CoinjoinError.BackendNotSynchronized);
+				}
 
-			if (WasabiBackendStatusProvide.LastResponse is not { } synchronizerResponse)
-			{
-				ScheduleRestartAutomatically(walletToStart, trackedAutoStarts, startCommand.StopWhenAllMixed, startCommand.OverridePlebStop, stoppingToken);
+				// If all coins are already private, then don't mix.
+				if (walletToStart.IsWalletPrivate() && startCommand.StopWhenAllMixed)
+				{
+					walletToStart.LogTrace("All mixed!");
 
-				NotifyCoinJoinStartError(walletToStart, CoinjoinError.BackendNotSynchronized);
-				return;
-			}
+					throw new CoinJoinClientException(CoinjoinError.AllCoinsPrivate);
+				}
 
-			// If all coins are already private, then don't mix.
-			if (await walletToStart.IsWalletPrivateAsync().ConfigureAwait(false))
-			{
-				walletToStart.LogTrace("All mixed!");
-				if (!startCommand.StopWhenAllMixed)
```

Comment:
Where did you move this condition?

Reply:
Moved up to LOC182 and removed the negation.

```cs
if (walletToStart.IsWalletPrivate() && startCommand.StopWhenAllMixed)
```

Reply:
You were right.

c2d7569 should put back the stopping condition